File: frontend/src/services/formula.ts
import api from '../config/api';

export interface Formula {
  id: number;
  name: string;
  description: string;
  category: string;
  version: string;
  path: string;
  repository: string;
  icon?: string;
  tags?: string[];
  metadata?: any;
  is_active: boolean;
}

export interface FormulaParameter {
  id: number;
  name: string;
  type: string;
  default?: any;
  required: boolean;
  label: string;
  description: string;
  validation?: any;
  order: number;
}

export interface FormulaTemplate {
  id: number;
  formula_id: number;
  name: string;
  description: string;
  pillar_data?: any;
  is_public: boolean;
  created_by: number;
  created_at: string;
  updated_at: string;
}

export interface FormulaDeployment {
  id: number;
  formula_id: number;
  name: string;
  description: string;
  target_hosts: string[];
  pillar_data?: any;
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';
  salt_job_id?: string;
  results?: any;
  success_count: number;
  failed_count: number;
  error_message?: string;
  started_by: number;
  started_at?: string;
  completed_at?: string;
  duration_seconds?: number;
  created_at: string;
  updated_at: string;
}

export interface ListFormulasResponse {
  formulas: Formula[];
  total: number;
  page: number;
  page_size: number;
}

export interface ListDeploymentsResponse {
  deployments: FormulaDeployment[];
  total: number;
  page: number;
  page_size: number;
}

export interface CreateTemplateRequest {
  name: string;
  description: string;
  pillar_data?: any;
  is_public?: boolean;
}

export interface CreateDeploymentRequest {
  formula_id: number;
  name: string;
  description?: string;
  target_hosts: string[];
  pillar_data?: any;
}

export interface CreateFormulaTemplateRequest {
  formula_id: number;
  name: string;
  description?: string;
  pillar_data?: any;
  is_public?: boolean;
}

export interface FormulaRepository {
  id: number;
  name: string;
  url: string;
  branch: string;
  local_path?: string;
  is_active: boolean;
  last_sync_at?: string;
  created_by: number;
  created_at: string;
  updated_at: string;
}

const formulaService = {
  // Formula管理
  listFormulas: async (page: number = 1, pageSize: number = 20, filters?: any): Promise<ListFormulasResponse> => {
    const response = await api.get('/formulas', { params: { page, page_size: pageSize, ...filters } });
    return response.data;
  },

  getFormula: async (id: number): Promise<{ formula: Formula; parameters: FormulaParameter[] }> => {
    const response = await api.get(`/formulas/${id}`);
    return response.data;
  },

  createFormula: async (formula: Partial<Formula>): Promise<Formula> => {
    const response = await api.post('/formulas', formula);
    return response.data.formula;
  },

  // Formula参数管理
  updateFormulaParameters: async (formulaId: number, parameters: FormulaParameter[]): Promise<void> => {
    await api.put(`/formulas/${formulaId}/parameters`, { parameters });
  },

  // Formula模板管理
  getFormulaTemplates: async (formulaId: number): Promise<{ templates: FormulaTemplate[] }> => {
    const response = await api.get(`/formulas/${formulaId}/templates`);
    return response.data;
  },

  createTemplate: async (formulaId: number, template: CreateTemplateRequest): Promise<FormulaTemplate> => {
    const response = await api.post(`/formulas/${formulaId}/templates`, template);
    return response.data.template;
  },

  createFormulaTemplate: async (template: CreateFormulaTemplateRequest): Promise<FormulaTemplate> => {
    const response = await api.post(`/formulas/${template.formula_id}/templates`, template);
    return response.data.template;
  },

  // Formula部署
  createDeployment: async (deployment: CreateDeploymentRequest): Promise<FormulaDeployment> => {
    const response = await api.post('/formulas/deployments', deployment);
    return response.data.deployment;
  },

  executeDeployment: async (deploymentId: number): Promise<void> => {
    await api.post(`/formulas/deployments/${deploymentId}/execute`);
  },

  listDeployments: async (page: number = 1, pageSize: number = 20, filters?: any): Promise<ListDeploymentsResponse> => {
    const response = await api.get('/formulas/deployments', { params: { page, page_size: pageSize, ...filters } });
    return response.data;
  },

  getDeployment: async (id: number): Promise<FormulaDeployment> => {
    const response = await api.get(`/formulas/deployments/${id}`);
    return response.data.deployment;
  },

  cancelDeployment: async (deploymentId: number): Promise<void> => {
    await api.post(`/formulas/deployments/${deploymentId}/cancel`);
  },

  cleanupOldDeployments: async (): Promise<{ message: string; count: number }> => {
    const response = await api.post('/formulas/deployments/cleanup');
    return response.data;
  },

  // Formula仓库管理
  createRepository: async (repo: any): Promise<any> => {
    const response = await api.post('/formulas/repositories', repo);
    return response.data.repository;
  },

  getRepository: async (id: number): Promise<FormulaRepository> => {
    const response = await api.get(`/formulas/repositories/${id}`);
    return response.data.repository;
  },

  updateRepository: async (id: number, repo: Partial<FormulaRepository>): Promise<FormulaRepository> => {
    const response = await api.put(`/formulas/repositories/${id}`, repo);
    return response.data.repository;
  },

  listRepositories: async (page: number = 1, pageSize: number = 20, filters?: any): Promise<any> => {
    const response = await api.get('/formulas/repositories', { params: { page, page_size: pageSize, ...filters } });
    return response.data;
  },

  syncRepository: async (repoId: number): Promise<void> => {
    await api.post(`/formulas/repositories/${repoId}/sync`);
  },
};

export default formulaService;

File: backend/internal/models/formula.go
package models

import (
	"time"

	"gorm.io/datatypes"
)

// Formula Salt Formula 模型
type Formula struct {
	ID          uint           `json:"id" gorm:"primarykey"`
	Name        string         `json:"name" gorm:"uniqueIndex;not null"`        // Formula 名称（如 nginx, redis）
	Description string         `json:"description"`                             // Formula 描述
	Category    string         `json:"category" gorm:"index"`                  // 分类（base, middleware, runtime, app）
	Version     string         `json:"version"`                                 // 版本号
	Path        string         `json:"path" gorm:"not null"`                   // 在仓库中的路径
	Repository  string         `json:"repository" gorm:"not null"`             // 仓库URL或本地路径
	Icon        string         `json:"icon"`                                   // 图标
	Tags        datatypes.JSON  `json:"tags" gorm:"type:jsonb"`                // 标签数组
	Metadata    datatypes.JSON  `json:"metadata" gorm:"type:jsonb"`            // 元数据（pillar schema, dependencies等）
	IsActive    bool           `json:"is_active" gorm:"default:true"`          // 是否启用
	CreatedBy   uint64         `json:"created_by" gorm:"not null;default:1"`   // 创建者ID，默认值为1
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   *time.Time     `json:"deleted_at,omitempty" gorm:"index"`
}

// FormulaParameter Formula 参数定义
type FormulaParameter struct {
	ID          uint           `json:"id" gorm:"primarykey"`
	FormulaID   uint           `json:"formula_id" gorm:"not null;index"`
	Name        string         `json:"name" gorm:"not null"`                              // 参数名称
	Type        string         `json:"type" gorm:"not null"`                              // 参数类型 (string, number, boolean, array, object)
	Default     datatypes.JSON `json:"default" gorm:"column:default_value;type:jsonb"`    // 默认值
	Required    bool           `json:"required" gorm:"default:false"`                     // 是否必填
	Label       string         `json:"label"`                                             // 显示标签
	Description string         `json:"description"`                                       // 参数描述
	Validation  datatypes.JSON `json:"validation" gorm:"type:jsonb"`                      // 验证规则
	Order       int            `json:"order" gorm:"column:order_index;default:0"`         // 显示顺序
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
}

// FormulaDeployment Formula 部署记录
type FormulaDeployment struct {
	ID             uint           `json:"id" gorm:"primarykey"`
	FormulaID      uint           `json:"formula_id" gorm:"not null;index"`
	Name           string         `json:"name" gorm:"not null"`                // 部署名称
	Description    string         `json:"description"`                         // 部署描述
	TargetHosts    datatypes.JSON `json:"target_hosts" gorm:"type:jsonb"`      // 目标主机列表
	PillarData     datatypes.JSON `json:"pillar_data" gorm:"type:jsonb"`       // Pillar 参数数据
	Status         string         `json:"status" gorm:"default:pending"`        // 状态 (pending, running, completed, failed, cancelled)
	SaltJobID      string         `json:"salt_job_id"`                         // Salt Job ID
	Results        datatypes.JSON `json:"results" gorm:"type:jsonb"`           // 执行结果
	SuccessCount   int            `json:"success_count" gorm:"default:0"`      // 成功主机数
	FailedCount    int            `json:"failed_count" gorm:"default:0"`       // 失败主机数
	ErrorMessage   string         `json:"error_message"`                       // 错误信息
	StartedBy      uint64         `json:"started_by" gorm:"not null;default:1"`  // 执行者ID
	StartedAt      *time.Time     `json:"started_at"`                          // 开始时间
	CompletedAt    *time.Time     `json:"completed_at"`                       // 完成时间
	Duration       int            `json:"duration_seconds"`                    // 执行时长(秒)
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
	DeletedAt      *time.Time     `json:"deleted_at,omitempty" gorm:"index"`
}

// FormulaTemplate Formula 模板（预设配置）
type FormulaTemplate struct {
	ID          uint           `json:"id" gorm:"primarykey"`
	FormulaID   uint           `json:"formula_id" gorm:"not null;index"`
	Name        string         `json:"name" gorm:"not null"`                   // 模板名称
	Description string         `json:"description"`                            // 模板描述
	PillarData  datatypes.JSON `json:"pillar_data" gorm:"type:jsonb"`         // 预设的Pillar数据
	IsPublic    bool           `json:"is_public" gorm:"default:false"`         // 是否公开
	CreatedBy   uint64         `json:"created_by" gorm:"not null;default:1"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   *time.Time     `json:"deleted_at,omitempty" gorm:"index"`
}

// FormulaRepository Formula 仓库配置
type FormulaRepository struct {
	ID          uint      `json:"id" gorm:"primarykey"`
	Name        string    `json:"name" gorm:"uniqueIndex;not null"`           // 仓库名称
	URL         string    `json:"url" gorm:"not null"`                        // Git仓库URL
	Branch      string    `json:"branch" gorm:"default:master"`               // 分支
	LocalPath   string    `json:"local_path"`                                 // 本地路径（如果不是Git仓库）
	IsActive    bool      `json:"is_active" gorm:"default:true"`              // 是否启用
	LastSyncAt  *time.Time `json:"last_sync_at"`                              // 最后同步时间
	CreatedBy   uint64    `json:"created_by" gorm:"not null;default:1"`       // 创建者ID，默认值为1
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
	DeletedAt   *time.Time `json:"deleted_at,omitempty" gorm:"index"`
}

// TableName 指定表名
func (Formula) TableName() string {
	return "formulas"
}

func (FormulaParameter) TableName() string {
	return "formula_parameters"
}

func (FormulaDeployment) TableName() string {
	return "formula_deployments"
}

func (FormulaTemplate) TableName() string {
	return "formula_templates"
}

func (FormulaRepository) TableName() string {
	return "formula_repositories"
}
